cdap-security/src/main/java/co/cask/cdap/security/store/FileSecureStore.java

```diff
@@ -0,0 +1,333 @@
+    } catch (NoSuchAlgorithmException | CertificateException e) {
+      throw new IOException("Unable to load the Secure Store. ", e);
+    }
+  }
+
+  private static boolean isBadOrWrongPassword(IOException ioe) {
+    /* According to Java keystore documentation if the load failed due to bad password
+       then the cause of the exception would be set to "UnrecoverableKeyException".
+       Unfortunately that is not the observed behavior. */
+    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+      return true;
+    }
+    // Workaround
+    return (ioe.getCause() == null)
+      && (ioe.getMessage() != null)
+      && ((ioe.getMessage().contains("Keystore was tampered")) || (ioe
+      .getMessage().contains("password was incorrect")));
+  }
+
+  /**
+   * Initialize the keyStore.
+   *
+   * @throws IOException If there is a problem reading or creating the keystore.
+   */
+  private static KeyStore locateKeystore(Path path, final char[] password) throws IOException {
+    Path newPath = constructNewPath(path);
+    KeyStore ks;
+    try {
+      ks = KeyStore.getInstance(SCHEME_NAME);
+      Files.deleteIfExists(newPath);
+      if (Files.exists(path)) {
+        // If the main file exists then the new path should not exist.
+        // Both existing means there is an inconsistency.
+        loadFromPath(ks, path, password);
+      } else {
+        // We were not able to load an existing key store. Create a new one.
+        ks.load(null, password);
+        LOG.info("New Secure Store initialized successfully.");
+      }
+    } catch (KeyStoreException | CertificateException | NoSuchAlgorithmException e) {
+      throw new IOException("Can't create Secure Store. ", e);
+    }
+    return ks;
+  }
+
+  /**
+   * Persist the keystore on the file system.
+   *
+   * During the flush the steps are
+   *  1. Delete the _NEW file if it exists, it will exist only if something had failed in the last run.
+   *  2. Try to write the current keystore in a _NEW file.
+   *  3. If something fails then revert the key store to the old state and throw IOException.
+   *  4. If everything is OK then rename the _NEW to the main file.
+   *
+   */
+  private void flush() throws IOException {
+    Path newPath = constructNewPath(path);
+    writeLock.lock();
+    try {
+      // Might exist if a backup has been restored etc.
+      Files.deleteIfExists(newPath);
+
+      // Flush the keystore, write the _NEW file first
+      writeToNew(newPath);
+      // Do Atomic rename _NEW to CURRENT
+      Files.move(newPath, path, ATOMIC_MOVE);
+    } catch (IOException ioe) {
+      throw ioe;
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  private void writeToNew(Path newPath) throws IOException {
+    try (OutputStream fos = new DataOutputStream(Files.newOutputStream(newPath))) {
+      keyStore.store(fos, password);
+    } catch (KeyStoreException e) {
+      throw new IOException("Can't store keystore " + this, e);
```

Comment:
This class doesn't implements `toString`, so the `+ this` might not be given useful info?